### ISSUE TITLE
consolidate query _name and boost support in query DSL

### ISSUE THREAD

**Issue by javanna** [elastic/elasticsearch#11744]
As part of the query refactoring we are trying to make support for `boost` and `_name` in our queries more generic and consistent (#10776) as it is currently copy pasted in all query parsers and builders, sometimes forgotten, and error prone. While working on that, I realized that `_name` and `boost` are supported slightly differently in the json depending on the type query. The main difference, besides bugs, is that some queries support both within the top level query object:

```
{
  "multi_match" : {
    "query" : "test",
    "fields" : ["field1", "field2"],
    "_name" : "query_name",
    "boost" : 10
  }
}
```

while others support them within the inner object named like the field that gets queried:

```
{
  "term" : {
    "field_name" : {
      "value" : "test",
      "_name" : "query_name",
      "boost" : 10
    }
  }
}
```

or even the following:

```
{
  "term" : {
    "field_name" : "value" : "test",
    "_name" : "query_name",
    "boost" : 10
  }
}
```

The following is a summary of all of the queries and their behaviour:

| Query                  | _name top level | _name inner object | boost top level | boost inner object |
|------------------------|:---------------:|:-----------------:|:---------------:|:-----------------:|
| and                    |        X        |                   |                 |                   |
| bool                   |        X        |                   |       X          |                   |
| boosting               |                 |                   |         X       |                   |
| common_terms        |                 |         X         |                 |        X          |
| constant_score         |                 |                   |         X       |                   |
| dis_max                |        X        |                   |         X       |                   |
| exists                 |        X        |                   |                 |                   |
| field_masking_span     |        X        |                   |         X       |                   |
| filtered               |        X        |                   |         X       |                   |
| function_score         |                 |                   |         X       |                   |
| fuzzy               |                 |         X         |                 |        X          |
| geo_bbox **           |        X        |                   |                 |                   |
| geo_distance **       |        X        |                   |                 |                   |
| geo_distance_range ** |        X        |                   |                 |                   |
| geohash_cell           |                 |                   |                 |                  |
| geo_polygon **        |        X        |                   |                 |                   |
| geo_shape **          |        X        |                   |         X       |                   |
| has_child              |        X        |                   |         X       |                   |
| has_parent             |        X        |                   |         X       |                   |
| ids                    |        X        |                   |         X       |                   |
| indices                |        X        |                   |                 |                   |
| limit                  |                 |                   |                 |                   |
| match_all              |                 |                   |         X       |                   |
| match               |                 |         X         |                 |          X        |
| missing                |        X        |                   |         X       |                   |
| more_like_this         |        X        |                   |         X       |                   |
| multi_match            |        X        |                   |         X       |                   |
| nested                 |        X        |                   |         X       |                   |
| not                    |        X        |                   |                 |                   |
| or                     |        X        |                   |                 |                   |
| prefix **          |        X        |         X         |                 |          X        |
| query_string           |        X        |                   |         X       |                   |
| range **           |                 |         X         |         X       |                   |
| regexp **          |        X        |         X         |                 |          X        |
| script                 |        X        |                   |                 |                   |
| simple_query_string    |        X        |                   |                 |                   |
| span_containing        |        X        |                   |         X       |                   |
| span_first             |        X        |                   |         X       |                   |
| span_multi             |                 |                   |                 |                   |
| span_near              |        X        |                   |         X       |                   |
| span_not               |        X        |                   |         X       |                   |
| span_or                |        X        |                   |         X       |                   |
| span_term           |                 |        X          |                 |         X         |
| span_within            |        X        |                   |         X       |                   |
| term **            |        X        |        X          |         X       |         X         |
| terms                  |        X        |                   |         X       |                   |
| type                   |                 |                   |                 |                   |
| wildcard            |                 |         X         |                 |          X        |

** : queries that support the inner object named as the field that gets queried, but support `_name` on the top level object instead.

All of the queries that support _name and boost within the inner object do so in their long version. The corresponding short version usually doesn't support  `_name` and `boost`, but only a single field in the form `field: value`.

1) The main question is where should the `_name` be supported. Should it always be in the top level object, as I was expecting without looking at the code, or does it make sense to wrap everything in the inner object if present?

2) Queries that don't support _name at all: boosting, constant_score, function_score, geohash_cell, limit, match_all, span_multi, type. Support for `_name` needs to be added to these?

3) Queries that don't support boost: indices, simple_query_string (PR opened), span_multi. Support for boost should be added to these?

4) Filters that don't support boost: and, bool, exists, geo_bbox, geo_distance, geo_distance_range, geohash_cell, geo_polygon, limit, missing, not, or, script, type. Now that filters and queries are merged, can these be used as queries? If so would the `boost` make sense if it was supported here too?

5) The following are inconsistencies that surely need to be fixed, according also to outcome of this discussion:
- ~~`GeoShapeQueryBuilder` doesn't support `boost`, but `GeoShapeQueryParser` parses it (fixed in #11810)~~.
- `PrefixQueryParser` supports `_name` in both short and long version, `boost` only in long version.
- `RegexpQueryParser` supports `_name` in both short and long version, `boost` only in long version.
- `RangeQueryParser` supports` _name` in the top level but `boost` in the inner field object.
- `TermQueryParser` supports both  `_name` and `boost` in both long and short version.
- ~~`TermsQueryBuilder` doesn't support `boost`, but parser parses it  (fixed in #11810)~~.

**Comment by jpountz** [comment]
> 1) The main question is where should the _name be supported. Should it always be in the top level object, as I was expecting without looking at the code, or does it make sense to wrap everything in the inner object if present?

+1 on only supporting it at the top level,  like our documentation describes.

> 2) Queries that don't support _name at all: boosting, constant_score, function_score, geohash_cell, limit, match_all, span_multi, type. Support for _name needs to be added to these?

Yes.

> 3) Queries that don't support boost: indices, simple_query_string (PR opened), span_multi. Support for boost should be added to these?

+1 on supporting `_boost` on all queries. I'm a bit torn about `indices` however which is not a real query. But on the other hand it would be easy to propagate the boos to the underlying query.

> 4) Filters that don't support boost: and, bool, exists, geo_bbox, geo_distance, geo_distance_range, geohash_cell, geo_polygon, limit, missing, not, or, script, type. Now that filters and queries are merged, can these be used as queries? If so would the boost make sense if it was supported here too?

Yes they can be used as queries, and most of them actually return the value of the boost as a score (like match_all) so I think it makes sense to be consistent and add support for `_boost` to these queries.

**Comment by clintongormley** [comment]
> +1 on only supporting it at the top level, like our documentation describes.

I'm not sure about this.  Every other extra parameter is specified in the inner object. What makes `_name` special?  It also means that you can't query a field called `_name` (admittedly an edge case).

Our documentation actually shows two conflicting syntaxes: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-named-queries-and-filters.html#search-request-named-queries-and-filters

My feeling is that it should be moved into the inner object for all queries.  It'd be easy to throw an exception if two "field names" are present, and an easy change to make in the application.

**Comment by javanna** [comment]
> My feeling is that it should be moved into the inner object for all queries. 

I see your point, but the majority of queries don't have an inner object, which would mean leaving things pretty much as they are? I find the current situation a bit counter intuitive: if there is an inner object, that's where the `_name` goes, otherwise top level (besides bugs and inconsistencies that need to be fixed). The other way around, if we are up for a big change, not sure but lemme throw it on the table, maybe we shouldn't have an inner object at all and have `field: field_name` in all queries instead and _name could be moved sasfely to the top level?

**Comment by clintongormley** [comment]
Follow up: @javanna and I talked and decided:

* all queries have a long form
* some queries (single field leaf queries) also have a short form
* in the short form, the only top-level parameter should be the field name
* specifying `_name`, `boost`, and any other parameters should always require the long form

**Comment by javanna** [comment]
I decided to do most of the work associated with this issue in the query-refactoring branch, as we have better tools to test these changes there e.g. unit tests and much more coverage. Also, the change has some breaking aspects which are not a problem in the query-refactoring branch as there we will completely change the way queries are sent over the wire anyway: if a builder prints out a new field that wasn't supported before when sending a search request through java api, the corresponding parser from an older node might throw error as it doesn't support that new field in that position.

The bigger part of this change was done as part of #11974. Besides what we have left to do is remove/deprecate support for `_name` and `boost` where they are supported but they shouldn't be:

- don't parse `_name` in short version of prefix query
- don't parse `_name` in short version of regexp query
- don't parse `_name` in the top level object of a range query (_name and boost should only be parsed in the inner field)
- don't parse `_name` and `boost` in short version of term query, only long version should support those

**Comment by clintongormley** [comment]
Closed by #11974

**Comment by javanna** [comment]
sorry @clintongormley I still have a couple of things to do so I can close this :) the four changes left listed above. will reopen and work on this.
